File: src/order.py
from api_calls import *

class Helper: 
    def Meter2Mile(Meter):
        return Meter / 1609


class Shop: 
    def __init__(self, Shop_Name, Postcode):
        self.Shop_Name = Shop_Name
        self.Postcode = Postcode

class Shops: 
    All = []
    TUN1 = Shop("TUN1", "ST75XX")
    TUN2 = Shop("TUN2", "ST34EE")
    All = [TUN1, TUN2]

class Order:
    def __init__(self, OrderNo, ItemString, Tel, Postcode):
        self.OrderNo = OrderNo 
        self.ItemString = ItemString 
        self.Tel = Tel
        self.Postcode = Postcode
        LatLong = []
        ShortestPath = ""

class OrderList:
    m = APICalls()
    def __init__(self, FromShop, OrderList=[]):
        self.FromShop = FromShop
        self.OrderList = OrderList
        
    def AddToList(self, OrderNo, ItemString, Tel, Postcode):
        order = Order(OrderNo, ItemString, Tel, Postcode)
        res = self.m.Get_LatLong(order.Postcode)
        order.LatLong = res
        self.OrderList.append(order)

    def IsAccessed(self, Postcode, Visited):
        if (Postcode in Visited):
            return True
        else:
            return False

    def OptimiseJourney(self):
        distances = []
        unique_postcodes = []
        for item in self.OrderList:
            if item.Postcode not in unique_postcodes:
                unique_postcodes.append(item.Postcode)
            for other_item in self.OrderList:
                if (item.Postcode != other_item.Postcode):
                    temp = self.m.Get_DistanceAndDuration(item.Postcode, other_item.Postcode)[0]
                    distances.append([item.Postcode,other_item.Postcode,temp])
        rf = RouteFinder();
        return rf.ShortestPath(unique_postcodes, distances, unique_postcodes[0]);

class RouteFinder:
    def RemoveWithStartPoint(self, NodesAndVertices, StartPoint):
        res = []
        for item in NodesAndVertices: 
            if (item[0] != StartPoint and item[1] != StartPoint):
                res.append(item)
        return res; 

    def GetBest(self, ModNodesAndVertices):
        shortest = 999999999999999999999999999999999;
        best = []
        for item in ModNodesAndVertices:
            if (item[2] < shortest):
                best = item;
                shortest = float(item[2])
        return best
            
    def FindWithStartPoint(self, NodesAndVertices, Postcode):
        Accessible = []
        for item in NodesAndVertices:
            if (item[0] == Postcode):
                Accessible.append(item)
            else:
                pass
        return Accessible

    def ShortestPath(self, Nodes, NodesAndVerticies, StartPoint):
        route = []
        route.append(StartPoint)
        CurrentNode = StartPoint
        All = NodesAndVerticies;
        
        while(len(route) != len(Nodes)):
            # Get all of the accessible nodes
            out = self.FindWithStartPoint(All, CurrentNode)
            out = self.GetBest(out);
            All = self.RemoveWithStartPoint(All, CurrentNode)
            CurrentNode = out[1]
            route.append(out[1])
        print(route)
        return route

File: src/order.js
const APICalls = require( './api_calls.js');
const api_calls = new APICalls();

class Order {
    constructor(OrderNo, ItemString, Tel, Postcode)
    {
        this.OrderNo = OrderNo ;
        this.ItemString = ItemString ;
        this.Tel = Tel;
        this.Postcode = Postcode;
        var LatLong = [];
        var ShortestPath = "";
    }
}

class OrderList {
    
    constructor(FromShop, Order_List=[])
    {
        this.FromShop = FromShop;
        this.Order_List = Order_List;
        var visited = [];
    }

    Add(OrderNo, ItemString, Tel, Postcode)
    {
        var order = new Order(OrderNo, ItemString, Tel, Postcode);
        this.Order_List.push(order);
    }
    
    PostcodeExists(value) {
        for (var i = 0 ; i < this.Order_List.length; i++)
        {   
            if (this.Order_List[i].Postcode === value)
            {
                return true;
            }
        }
        return false;
    }

    async OptimiseJourney()
    {
        this.Add(-1,-1,-1,this.FromShop);
        var distances = [];
        var unique_postcodes = [];

        for (var x = 0; x < this.Order_List.length; x++)
        {
            if (this.PostcodeExists(this.Order_List[x].Postcode, this.Order_List))
            {
                unique_postcodes.push(this.Order_List[x].Postcode);
            }
        }
        
        for (var i = 0; i < this.Order_List.length; i++)
        {
            for (var j = 0 ; j < this.Order_List.length; j++)
            {
                if (!(this.Order_List[i].Postcode === this.Order_List[j].Postcode))
                {
                    var temp = await api_calls.Get_Distance(this.Order_List[i].Postcode, this.Order_List[j].Postcode);
                    distances.push([this.Order_List[i].Postcode, this.Order_List[j].Postcode, temp]);
                }
            }
        }
        var rf = new RouteFinder();
        var shortest = await rf.ShortestPath(unique_postcodes, distances, this.FromShop);
        console.log(shortest)
        var all_orders = [];
        for (var y = 0 ; y < shortest.length; y++)
        {
            for (var z = 0; z < this.Order_List.length; z++)
            {
                if (this.Order_List[z].Postcode === shortest[y])
                {
                    all_orders.push(this.Order_List[z]);
                    console.log (" - " +this.Order_List[z].Postcode + "===" + shortest[y]);
                }
            }
        }
        console.log(all_orders)
        return all_orders;
    }
}

class RouteFinder {
    RemoveWithStartPoint(NodesAndVertices, StartPoint)
    {
       var res = [];
       for (var i = 0 ; i < NodesAndVertices.length; i++)
       {
            if (NodesAndVertices[i][0] != StartPoint && NodesAndVertices[i][1] != StartPoint)
            {
                res.push(NodesAndVertices[i]);
            }
       }
       return res;
    }

    GetBest(ModNodesAndVertices)
    {
        var shortest = Number.MAX_SAFE_INTEGER;
        var best = [];
        for (var i = 0 ; i < ModNodesAndVertices.length; i++)
        {
            if (ModNodesAndVertices[i][2] < shortest)
            {
                best = ModNodesAndVertices[i];
                shortest = parseFloat(ModNodesAndVertices[i][2]);
            }
        }
        return best;
    }

    FindWithStartPoint(NodesAndVertices, Postcode)
    {
        var Accessible = [];
        for (var item = 0; item < NodesAndVertices.length; item++)
        {
            if (NodesAndVertices[item][0]==Postcode)
            {
                Accessible.push(NodesAndVertices[item]);
            }
        }
        return Accessible;
    }

    ShortestPath(Nodes, NodesAndVertices, StartPoint)
    {
        let route = [];
        route.push(StartPoint);
        let CurrentNode = StartPoint;
        let All = NodesAndVertices;
//

        while (route.length != Nodes.length)
        {
            let out = this.FindWithStartPoint(All, CurrentNode);
            out = this.GetBest(out);
            All = this.RemoveWithStartPoint(All, CurrentNode);
            CurrentNode = out[1];
            route.push(out[1]);
        }
        return route; 
    }
}

module.exports = {
    Order,
    OrderList,
    RouteFinder
  };